libcxx/test/support/constexpr_random.h

```diff
@@ -0,0 +1,438 @@
+#endif
+  template <class URNG>
+  TEST_CONSTEXPR_CXX14 result_type operator()(URNG& gen, const param_type& param)
+      TEST_DISABLE_UBSAN_UNSIGNED_INTEGER_CHECK {
+    static_assert(detail::is_valid_urng<URNG>::value, "invalid uniform random bit generator used");
+    typedef typename std::conditional<sizeof(result_type) <= sizeof(std::uint32_t),
+                                      std::uint32_t,
+                                      typename std::make_unsigned<result_type>::type>::type UIntType;
+    const UIntType rp = UIntType(param.b()) - UIntType(param.a()) + UIntType(1);
+    if (rp == 1)
+      return param.a();
+    const std::size_t ur_digits = std::numeric_limits<UIntType>::digits;
+    typedef detail::independent_bits_engine<URNG, UIntType> Eng;
+    if (rp == 0)
+      return static_cast<result_type>(Eng(gen, ur_digits)());
+    std::size_t width = ur_digits - detail::countl_zero(rp) - 1;
+    if ((rp & (std::numeric_limits<UIntType>::max() >> (ur_digits - width))) != 0)
+      ++width;
+    Eng eng(gen, width);
+    UIntType eng_result = 0;
+    do {
+      eng_result = eng();
+    } while (eng_result >= rp);
+    return static_cast<result_type>(eng_result + param.a());
+  }
+#undef TEST_DISABLE_UBSAN_UNSIGNED_INTEGER_CHECK
+
+  // property functions
+  TEST_CONSTEXPR result_type a() const { return param_.a(); }
+  TEST_CONSTEXPR result_type b() const { return param_.b(); }
+
+  TEST_CONSTEXPR param_type param() const { return param_; }
+  TEST_CONSTEXPR_CXX14 void param(const param_type& param) { param_ = param; }
+
+  TEST_CONSTEXPR result_type min() const { return a(); }
+  TEST_CONSTEXPR result_type max() const { return b(); }
+
+#if TEST_STD_VER >= 20
+  friend bool operator==(const uniform_int_distribution&, const uniform_int_distribution&) = default;
+#else
+  TEST_CONSTEXPR friend bool operator==(const uniform_int_distribution& lhs, const uniform_int_distribution& rhs) {
+    return lhs.param_ == rhs.param_;
+  }
+  TEST_CONSTEXPR friend bool operator!=(const uniform_int_distribution& lhs, const uniform_int_distribution& rhs) {
+    return !(lhs == rhs);
+  }
+#endif
+};
+
+class simple_random_generator {
+private:
+  std::uint32_t status_;
+
+public:
+  typedef std::uint16_t result_type;
+
+  static TEST_CONSTEXPR result_type min() TEST_NOEXCEPT { return 0; }
+  static TEST_CONSTEXPR result_type max() TEST_NOEXCEPT { return static_cast<result_type>(-1); }
+#if TEST_STD_VER < 11
+  static const result_type min_value = 0;
+  static const result_type max_value = static_cast<result_type>(-1);
+#endif
+  static TEST_CONSTEXPR const result_type default_seed = 19937;
+
+#if TEST_STD_VER >= 11
+  constexpr simple_random_generator() noexcept : simple_random_generator(default_seed) {}
+#else
+  simple_random_generator() throw() : status_(default_seed) {}
+#endif
+  TEST_CONSTEXPR explicit simple_random_generator(std::uint16_t s) TEST_NOEXCEPT : status_(s) {}
+
+  TEST_CONSTEXPR_CXX14 result_type operator()() TEST_NOEXCEPT {
```

Comment:
Can you add a comment which engine this is and which values values?
For example
`A linear congruential generator, using the constants used by MSVC.`

Reply:
Comment added - `// A linear congruential generator, using the constants used by MS UCRT.`